### PR TITLE
WiP Optimize the backend

### PR DESCRIPTION
**Pull request by sjrd**

### CODE REVIEW

File: ir/shared/src/main/scala/org/scalajs/ir/Position.scala [hunk 1]
```diff
@@ -20,16 +20,18 @@ final case class Position(
...
     def isEmptySlowPath(): Boolean = {
       source.getScheme == null && source.getRawAuthority == null &&
         source.getRawQuery == null && source.getRawFragment == null
     }
     source.getRawPath == "" && isEmptySlowPath()
   }
 
+  def show: String = s"$line:$column"
+
+  def isEmpty: Boolean = _isEmpty
```
Comment (gzm0) [inline ir/shared/src/main/scala/org/scalajs/ir/Position.scala:33]: Have you considered making `source` nullable?

**Comment by gzm0** [comment]
Just to make sure this is on your radar: PathOutputImpl reads the target file first to avoid spurious writes:
https://github.com/scala-js/scala-js/blob/f3099bafbc5f263d37102e6639f356f0309d5e0b/linker/jvm/src/main/scala/org/scalajs/linker/PathOutputDirectory.scala#L38-L43

We might be in a position now to avoid this, at least for incremental runs.

File: ir/shared/src/main/scala/org/scalajs/ir/Position.scala (hunk 1 above)
Comment (sjrd) [reply ir/shared/src/main/scala/org/scalajs/ir/Position.scala:33]: I wish I could do that, but `source` is a very much a public field, and I don't want to expose a nullable field in our public API.

**Comment by sjrd** [comment]
> We might be in a position now to avoid this, at least for incremental runs.

Yes, it's on my radar. After all the changes I've been doing all week, we're close to as good as it gets for the single-module case. But for the multi-module case, flame graphs blame all the useless reading I/O a lot, so there's definitely something to do there.

**Comment by sjrd** [comment]
Superseded. All the commits in here have been merged in one PR or another.

**Pull request closed without merging**